ios/MullvadVPN/AlertPresenter.swift

```diff
@@ -0,0 +1,97 @@
+//
+//  AlertPresenter.swift
+//  MullvadVPN
+//
+//  Created by pronebird on 04/06/2020.
+//  Copyright © 2020 Mullvad VPN AB. All rights reserved.
+//
+
+import Foundation
+import UIKit
+
+private let kUIAlertControllerDidDissmissNotification = Notification.Name("UIAlertControllerDidDismiss")
+
+class AlertPresenter {
+    private enum ExclusivityCategory {
+        case exclusive
+    }
+
+    private let operationQueue = OperationQueue()
+    private lazy var exclusivityController = ExclusivityController<ExclusivityCategory>(operationQueue: operationQueue)
+
+    init() {
+        _ = AlertPresenterUIKitHooks.once
+    }
+
+    func enqueue(_ alertController: UIAlertController, presentingController: UIViewController, presentCompletion: (() -> Void)? = nil) {
+        let operation = PresentAlertOperation(
+            alertController: alertController,
+            presentingController: presentingController,
+            presentCompletion: presentCompletion
+        )
+
+        exclusivityController.addOperation(operation, categories: [.exclusive])
+    }
+
+}
+
+private class PresentAlertOperation: AsyncOperation {
+    private let alertController: UIAlertController
+    private let presentingController: UIViewController
+    private var dismissalObserver: NSObjectProtocol?
+    private let presentCompletion: (() -> Void)?
+
+    init(alertController: UIAlertController, presentingController: UIViewController, presentCompletion: (() -> Void)? = nil) {
+        self.alertController = alertController
+        self.presentingController = presentingController
+        self.presentCompletion = presentCompletion
+
+        super.init()
+    }
+
+    override func main() {
+        DispatchQueue.main.async {
+            self.dismissalObserver = NotificationCenter.default.addObserver(
+                forName: kUIAlertControllerDidDissmissNotification,
+                object: self.alertController,
+                queue: nil,
+                using: { [weak self] (note) in
+                    self?.finish()
+            })
+
+            self.presentingController.present(self.alertController, animated: true, completion: self.presentCompletion)
+        }
+    }
+}
+
+/// A helper struct that swizzles `viewDidDisappear` on `UIAlertController` in order to be able to
+/// detect when the controller disappears.
+/// The event is broadcasted via `kUIAlertControllerDidDissmissNotification` notification.
+private struct AlertPresenterUIKitHooks {
```

Comment:
This is a patch on UIKit to be able to intercept the `viewDidDisapper` on `UIAlertController` and submit `Notification` to tell `AlertPresenter` when `UIAlertController` disappears. This is needed because only one `UIAlertController` can be presented at a time and there is no easy way to know when it's being dismissed. This is nasty but hopefully shouldn't cause much trouble.